### PR TITLE
feat: throw condition exception on TestNode exception

### LINKED ISSUE #472: Exceptions thrown in the rules engine during propagate-items-to-nodes need better debugging information
When an exception is thrown inside the method `propagate-items-to-nodes` the stack trace provides very little details about which rule was it that threw the exception and where to handle. For this reason I think it would be much useful if an exception was thrown with cause and better debugging information such as the condition and rule (by fully qualified namespace name) where it was thrown.

Example new message:
```
Condition exception raised.
condition:
[?fact-binding <- TemperatureControlBlah]
rules:
my-rules.namespace.core/my-rule
```

java.lang.NullPointerException: null
 at clojure.lang.Numbers.ops (Numbers.java:1068)
    clojure.lang.Numbers.gte (Numbers.java:263)
    clojure.core$eval4653$TN_2647_TE__5713.invoke (NO_SOURCE_FILE:0)
    clojure.core$filter$fn__5911.invoke (core.clj:2821)
    clojure.lang.LazySeq.sval (LazySeq.java:42)
    clojure.lang.LazySeq.seq (LazySeq.java:51)
    clojure.lang.RT.seq (RT.java:535)
    clojure.core$seq__5419.invokeStatic (core.clj:139)
    clojure.core/seq (core.clj:139)
    clara.rules.engine$propagate_items_to_nodes.invokeStatic (engine.cljc:164)
    clara.rules.engine$propagate_items_to_nodes.invoke (engine.cljc:147)
    clara.rules.engine.LocalTransport.send_tokens (engine.cljc:179)
    clara.rules.engine.TestNode.left_activate (engine.cljc:940)
    clara.rules.engine$fn__12690$G__12637__12697.invoke (engine.cljc:99)
    clara.rules.engine$fn__12690$G__12636__12705.invoke (engine.cljc:99)
    clara.rules.engine$propagate_items_to_nodes.invokeStatic (engine.cljc:165)
    clara.rules.engine$propagate_items_to_nodes.invoke (engine.cljc:147)
    clara.rules.engine.LocalTransport.send_tokens (engine.cljc:179)
    clara.rules.engine$send_accumulated.invokeStatic (engine.cljc:990)
    clara.rules.engine$send_accumulated.invoke (engine.cljc:970)
    clara.rules.engine.AccumulateNode.left_activate (engine.cljc:1058)
    clara.rules.engine$fn__12690$G__12637__12697.invoke (engine.cljc:99)
    clara.rules.engine$fn__12690$G__12636__12705.invoke (engine.cljc:99)
    clara.rules.engine$propagate_items_to_nodes.invokeStatic (engine.cljc:155)
    clara.rules.engine$propagate_items_to_nodes.invoke (engine.cljc:147)
    clara.rules.engine.LocalTransport.send_tokens (engine.cljc:179)
    clara.rules.engine$send_accumulated.invokeStatic (engine.cljc:990)
    clara.rules.engine$send_accumulated.invoke (engine.cljc:970)
    clara.rules.engine.AccumulateNode.left_activate (engine.cljc:1036)
    clara.rules.engine$fn__12690$G__12637__12697.invoke (engine.cljc:99)
    clara.rules.engine$fn__12690$G__12636__12705.invoke (engine.cljc:99)
    clara.rules.engine$propagate_items_to_nodes.invokeStatic (engine.cljc:165)
    clara.rules.engine$propagate_items_to_nodes.invoke (engine.cljc:147)
    clara.rules.engine.LocalTransport.send_tokens (engine.cljc:179)
    clara.rules.engine$send_accumulated.invokeStatic (engine.cljc:990)
    clara.rules.engine$send_accumulated.invoke (engine.cljc:970)
    clara.rules.engine.AccumulateNode.left_activate (engine.cljc:1058)
    clara.rules.engine$fn__12690$G__12637__12697.invoke (engine.cljc:99)
    clara.rules.engine$fn__12690$G__12636__12705.invoke (engine.cljc:99)
    clara.rules.engine$propagate_items_to_nodes.invokeStatic (engine.cljc:155)
    clara.rules.engine$propagate_items_to_nodes.invoke (engine.cljc:147)
    clara.rules.engine.LocalTransport.send_tokens (engine.cljc:179)
    clara.rules.engine$send_accumulated.invokeStatic (engine.cljc:990)
    clara.rules.engine$send_accumulated.invoke (engine.cljc:970)
    clara.rules.engine.AccumulateWithJoinFilterNode.left_activate (engine.cljc:1404)
    clara.rules.engine$fn__12690$G__12637__12697.invoke (engine.cljc:99)
    clara.rules.engine$fn__12690$G__12636__12705.invoke (engine.cljc:99)
    clara.rules.engine$propagate_items_to_nodes.invokeStatic (engine.cljc:155)
    clara.rules.engine$propagate_items_to_nodes.invoke (engine.cljc:147)
    clara.rules.engine.LocalTransport.send_tokens (engine.cljc:179)
    clara.rules.engine.RootJoinNode.right_activate (engine.cljc:582)
    clara.rules.engine$fn__12718$G__12712__12725.invoke (engine.cljc:107)
    clara.rules.engine$fn__12718$G__12711__12733.invoke (engine.cljc:107)
    clara.rules.engine$propagate_items_to_nodes.invokeStatic (engine.cljc:165)
    clara.rules.engine$propagate_items_to_nodes.invoke (engine.cljc:147)
    clara.rules.engine.LocalTransport.send_elements (engine.cljc:176)
    clara.rules.engine.AlphaNode.alpha_activate (engine.cljc:538)
    clara.rules.engine$flush_updates$flush_all__13057.invoke (engine.cljc:269)
    clara.rules.engine$flush_updates.invokeStatic (engine.cljc:277)
    clara.rules.engine$flush_updates.invoke (engine.cljc:252)
    clara.rules.engine$fire_rules_STAR_.invokeStatic (engine.cljc:1839)
    clara.rules.engine$fire_rules_STAR_.invoke (engine.cljc:1732)
    clara.rules.engine.LocalSession.fire_rules (engine.cljc:1925)
    clara.rules$fire_rules.invokeStatic (rules.cljc:44)
    clara.rules$fire_rules.invoke (rules.cljc:29)
    rules.engine$run_rules_BANG_.invokeStatic (engine.clj:72)
    rules.engine$run_rules_BANG_.doInvoke (engine.clj:54)
    clojure.lang.RestFn.applyTo (RestFn.java:139)
    clojure.core$apply.invokeStatic (core.clj:669)
    clojure.core$apply.invoke (core.clj:662)
    rules.engine$run_rules_async_BANG_$fn__17913.invoke (engine.clj:139)
    async_util.pool$future_supplier$reify__16298.get (pool.clj:31)
    java.util.concurrent.CompletableFuture$AsyncSupply.run (:-1)
    java.util.concurrent.ThreadPoolExecutor.runWorker (:-1)
    java.util.concurrent.ThreadPoolExecutor$Worker.run (:-1)
    java.lang.Thread.run (:-1)

### PR DESCRIPTION
**Pull request by k13gomez**
ensure we throw a more detailed condition exception when an exception is encountered during the evaluation of a TestNode expression

Closes #472

### CODE REVIEW

**Comment by k13gomez** [comment]
I need to add some test cases, but working on this feature so that we get more detailed exceptions (at a minimum knowing the rule where the exception is thrown)

**Comment by EthanEChristian** [comment]
@k13gomez,

Would it be possible to log an issue for this feature, and provide a sample of the current exception and an ideal message?

**Comment by k13gomez** [comment]
I created issue #472 with some details, also I will be adding unit tests for clarity.

**Comment by k13gomez** [comment]
New stacktrace printout:
clojure.lang.ExceptionInfo: Condition exception raised.
when processing fact
  #clara.rules.testfacts.WindSpeed{:windspeed 10, :location "MCI"}
  #clara.rules.testfacts.Temperature{:temperature nil, :location "MCI"}
with bindings
  {:?w 10, :?t nil}
Conditions:

1. [:test (> ?t ?w)]
   queries:
     my-test-query

 at clojure.core$ex_info.invoke (core.clj:4593)
    clara.rules.engine$throw_condition_exception.invoke (engine.cljc:507)
    clara.rules.engine$test_node_matches$fn__6131.invoke (engine.cljc:934)
    clara.rules.engine$test_node_matches.invoke (engine.cljc:931)
    clojure.core$partial$fn__4531.invoke (core.clj:2507)
    clojure.core$filter$fn__4580.invoke (core.clj:2690)
    clojure.lang.LazySeq.sval (LazySeq.java:40)
    clojure.lang.LazySeq.seq (LazySeq.java:49)
    clojure.lang.RT.seq (RT.java:507)
    clojure.core/seq (core.clj:137)
    clara.rules.engine$propagate_items_to_nodes.invoke (engine.cljc:164)
    clara.rules.engine.LocalTransport.send_tokens (engine.cljc:179)
    clara.rules.engine.TestNode.left_activate (engine.cljc:953)
    clara.rules.engine$eval4927$fn__4971$G__4913__4978.invoke (engine.cljc:99)
    clara.rules.engine$eval4927$fn__4971$G__4912__4986.invoke (engine.cljc:99)
    clara.rules.engine$propagate_items_to_nodes.invoke (engine.cljc:165)
    clara.rules.engine.LocalTransport.send_tokens (engine.cljc:179)
    clara.rules.engine.HashJoinNode.right_activate (engine.cljc:651)
    clara.rules.engine$eval5008$fn__5009$G__4998__5016.invoke (engine.cljc:107)
    clara.rules.engine$eval5008$fn__5009$G__4997__5024.invoke (engine.cljc:107)
    clara.rules.engine$propagate_items_to_nodes.invoke (engine.cljc:165)
    clara.rules.engine.LocalTransport.send_elements (engine.cljc:176)
    clara.rules.engine.AlphaNode.alpha_activate (engine.cljc:540)
    clara.rules.engine.LocalSession/fn (engine.cljc:1932)
    clara.rules.engine.LocalSession.fire_rules (engine.cljc:1925)
    clara.rules$fire_rules.invoke (rules.cljc:44)
    clara.test_rules$eval13619$fn__13620.invoke (form-init8953837220311696004.clj:2376)
    clara.test_rules$eval13619.invoke (form-init8953837220311696004.clj:2373)
    clojure.lang.Compiler.eval (Compiler.java:6782)
    clojure.lang.Compiler.eval (Compiler.java:6745)
    clojure.core$eval.invoke (core.clj:3081)
    nrepl.middleware.interruptible_eval$evaluate$fn__744.invoke (interruptible_eval.clj:91)
    clojure.main$repl$read_eval_print__7099$fn__7102.invoke (main.clj:240)
    clojure.main$repl$read_eval_print__7099.invoke (main.clj:240)
    clojure.main$repl$fn__7108.invoke (main.clj:258)
    clojure.main$repl.doInvoke (main.clj:258)
    clojure.lang.RestFn.invoke (RestFn.java:1523)
    nrepl.middleware.interruptible_eval$evaluate.invoke (interruptible_eval.clj:84)
    nrepl.middleware.interruptible_eval$interruptible_eval$fn__770$fn__774.invoke (interruptible_eval.clj:155)
    clojure.lang.AFn.run (AFn.java:22)
    nrepl.middleware.session$session_exec$main_loop__867$fn__869.invoke (session.clj:190)
    nrepl.middleware.session$session_exec$main_loop__867.invoke (session.clj:189)
    clojure.lang.AFn.run (AFn.java:22)
    java.lang.Thread.run (Thread.java:748)

File: src/main/clojure/clara/rules/engine.cljc [hunk 1]
```diff
@@ -937,15 +950,19 @@
...
     (retract-tokens transport memory listener children tokens))
 
   (get-join-keys [node] [])
 
-  (description [node] (str "TestNode -- " (:text test))))
+  (description [node] (str "TestNode -- " (:text test)))
+
+  IConditionNode
+  (get-condition-description [this]
+    (into [:test] (:constraints (meta test)))))
```
Comment (EthanEChristian) [inline src/main/clojure/clara/rules/engine.cljc:965]: Might be interesting to have a test that confirms that this metadata is maintained across a serialization/deserialization. 

I feel fairly confident that the metadata would survive as the entire form should be captured, but incase something changes in the future, a test to catch it would be good.

The test would likely need to be added somewhere in the durability test:
https://github.com/cerner/clara-rules/blob/main/src/test/clojure/clara/test_durability.clj

**Comment by EthanEChristian** [comment]
Overall, i think this makes sense and I am glad you were able to thread it into the pattern that we have for other node types.

File: src/main/clojure/clara/rules/engine.cljc (hunk 1 above)
Comment (k13gomez) [reply src/main/clojure/clara/rules/engine.cljc:965]: Sure, I will add that in as we also rely serialization heavily; I'm testing it live today on our backend.

Comment (k13gomez) [reply src/main/clojure/clara/rules/engine.cljc:965]: @EthanEChristian I've added a test to ensure that the metadata is maintained across a serialization/deserialization, it worked as expected. I also tested this on a live system and worked as expected.

**Comment by k13gomez** [comment]
If this looks good to merge can you approve? Ideally I'd like to get this in and cut a release.

File: src/main/clojure/clara/rules/compiler.clj [hunk 2]
```diff
@@ -415,17 +415,20 @@
...
         fn-name (mk-node-fn-name "TestNode" node-id "TE")]
 
-    `(fn ~fn-name [~'?__token__]
-       (let [~@assignments]
-
-        (and ~@tests)))))
+    `(let [handler# (fn ~fn-name [~'?__token__]
+                      (let [~@assignments]
+                        (and ~@constraints)))]
+       (with-meta handler#
```
Comment (mrrodriguez) [inline src/main/clojure/clara/rules/compiler.clj:428]: There may be a reason to be concerned here about the overhead of using `with-meta` to wrap the fn.

`with-meta` wraps the fn in a RestFn here https://github.com/clojure/clojure/blob/clojure-1.10.3/src/jvm/clojure/lang/AFunction.java#L29
which has the unfortunate consequence of making all calls to the wrapped fn go through variadic arg runtime unrolling via `apply` https://github.com/clojure/clojure/blob/clojure-1.10.3/src/jvm/clojure/lang/AFunction.java#L31

The constraint fn's themselves may be called in hot spots so this can be concerning.

Microbench mark example (ran it a bunch of times to stabilize the numbers):
```clj
(let [f (fn [x] (inc x))
      mf (with-meta f {:my :meta})]

  (time (dotimes [_ 1e5]
          (f 2)))

  (time (dotimes [_ 1e5]
          (mf 2))))
;;= "Elapsed time: 3.851125 msecs"
;;= "Elapsed time: 22.38039 msecs"
;;= nil
```

File: src/main/clojure/clara/rules/engine.cljc [hunk 3]
```diff
@@ -494,7 +494,9 @@
                                (str "with bindings\n  " bindings))
         message-header (string/join ["Condition exception raised.\n"
                                      "when processing fact\n"
-                                     (str "  " (pr-str fact) "\n")
+                                     (str "  " (->> (flatten [fact])
```
Comment (mrrodriguez) [inline src/main/clojure/clara/rules/engine.cljc:497]: As a general principle I think it almost always is a red flag to need to use `flatten`. What is arbitrary and out of control about the structure here that makes this necessary? 

If the problem is just that sometimes `fact` is not a fact, but rather a collection of facts, I think that should be handled explicitly to only  unroll that top level.

Better yet, I think the name should just be changed to `facts` and all callers should pass a collection. If they only have 1 they can put that in a coll too.

File: src/main/clojure/clara/rules/engine.cljc [hunk 4]
```diff
@@ -937,15 +950,19 @@
      memory
      listener
      children
-     (filter test tokens)))
+     (filter (partial test-node-matches node test {}) tokens)))
```
Comment (mrrodriguez) [inline src/main/clojure/clara/rules/engine.cljc:953]: FYI: `partial` degrades performance possibly significantly beyond it's 4 arity case. So this is pushing it right up against the limit.

File: src/test/clojure/clara/test_rules.clj [hunk 5]
```diff
@@ -2361,7 +2361,22 @@
                   (query check-exception))
               (is false "Scope binding test is expected to throw an exception.")
               (catch Exception e e))]
+      (is (= {:?w 10, :?t nil}
+             (-> e (ex-data) :bindings))))))
+
+(deftest test-local-scope-visible-in-test-filter
```
Comment (mrrodriguez) [inline src/test/clojure/clara/test_rules.clj:2367]: I'm confused by this test case's name since it seems to be checking that the `:test` node is throwing an exception due to the `?t` value being nil right?

File: src/test/clojure/clara/test_rules.clj [hunk 6]
```diff
@@ -2361,7 +2361,22 @@
...
+                                                    [:test (> ?t ?w)]])
+                               :name "my-test-query")]
 
+    (let [e (try
+              (-> (mk-session [check-exception])
+                  (insert (->WindSpeed 10 "MCI") (->Temperature nil "MCI"))
+                  (fire-rules)
+                  (query check-exception))
+              (is false "Scope binding test is expected to throw an exception.")
+              (catch Exception e e))]
```
Comment (mrrodriguez) [inline src/test/clojure/clara/test_rules.clj:2379]: Shouldn't there be some validation that the exception has the useful information that was added in this PR?

File: src/main/clojure/clara/rules/compiler.clj (hunk 2 above)
Comment (k13gomez) [reply src/main/clojure/clara/rules/compiler.clj:428]: True, there’s a bit of an overhead, the alternative approach I was considering was creating a protocol for the test expression handler instead of using a Fn directly so that it could carry the additional metadata needed for introspection at runtime, or just simply capturing that data when compile-test is called. It will require refactoring more tests but I can make an alternative PR with that approach.

Comment (mrrodriguez) [reply src/main/clojure/clara/rules/compiler.clj:428]: One possible way out of this sort of mess would be if this fn could just return this map instead and then 
https://github.com/cerner/clara-rules/blob/014ed43ed2960d64db065a1fd108dce336aac960/src/main/clojure/clara/rules/compiler.clj#L1443
would use the `:handler` value directly.

However, I think then there may be an issue with the durability layer persisting this meta information via serde...

**Review by mrrodriguez — COMMENTED** [review]
I have some comments and concerns. 

Also, I'm wondering if the PR title is correct. My understanding of this issue is that there is a desire to add better error reporting for `:test` nodes specifically correct?

File: src/main/clojure/clara/rules/engine.cljc (hunk 3 above)
Comment (mrrodriguez) [reply src/main/clojure/clara/rules/engine.cljc:497]: Also is there any chance that this list of facts could be very large and cause a flood of output, eg. from accumulated coll values?

**Comment by k13gomez** [comment]
Thanks for all the feedback, I’ll work on this over the next few weeks, I think some of the questions and concerns should also lead to some new test cases.

File: src/main/clojure/clara/rules/compiler.clj (hunk 2 above)
Comment (k13gomez) [reply src/main/clojure/clara/rules/compiler.clj:428]: @mrrodriguez in my experience it gets persisted just fine because the durability layer persists the productions but removes the compiled portion, so it would recompile them whenever the session was reassembled.

File: src/main/clojure/clara/rules/engine.cljc (hunk 3 above)
Comment (k13gomez) [reply src/main/clojure/clara/rules/engine.cljc:497]: @mrrodriguez I updated the PR title and description, I ended up rewriting the initial PR and targeting just the TestNode which is where there was missing exception handling logic; now, regarding the list of facts, this is the same possibility of flood of output with all the other Condition Exceptions where bindings are listed, because it is possible for the bindings used in a TestNode to be a list of accumulated facts all the same. I am wondering if I should change it so that the condition exception raised by the TestNode does not actually list facts because it does not really bind to any fact but it is purely dependent on the bindings and that is all that should be needed along with the condition/query encountering the exception.

File: src/main/clojure/clara/rules/engine.cljc (hunk 4 above)
Comment (k13gomez) [reply src/main/clojure/clara/rules/engine.cljc:953]: Good point, I'm going to remove the `env` argument since it doesn't really get used, I was just sticking to the existing pattern. Another possibility is to write this as an eager-for comprehension, so I'm going to give that a try.

File: src/test/clojure/clara/test_rules.clj (hunk 5 above)
Comment (k13gomez) [reply src/test/clojure/clara/test_rules.clj:2367]: I'll give it a more descriptive name and actually add additional assertions since there's a couple more things I want to check on this exception.

File: src/test/clojure/clara/test_rules.clj (hunk 6 above)
Comment (k13gomez) [reply src/test/clojure/clara/test_rules.clj:2379]: Yes, I'll add additional assertions to validate that.

File: src/test/clojure/clara/test_durability.clj [hunk 7]
```diff
@@ -338,6 +339,29 @@
...
+          rb (-> s eng/components :rulebase)
+          deserialized1 (rb-serde s nil)
+          ;; Need a session to do the 2nd round of SerDe.
+          restored1 (d/assemble-restored-session deserialized1 {})
+          deserialized2 (rb-serde restored1 nil)
+          restored2 (d/assemble-restored-session deserialized2 {})]
+      (is (= [[:test '(not-empty ?ts)]]
+             (map eng/get-condition-description (get-test-nodes s))
+             (map eng/get-condition-description (get-test-nodes restored1))
+             (map eng/get-condition-description (get-test-nodes restored2)))))))
```
[the remark this answers is not in the document]
Comment (k13gomez) [reply src/test/clojure/clara/test_durability.clj]: I've improved this test it now fires and then re-fires the serialized session as described above.

**Comment by k13gomez** [comment]
I've made several changes to address the PR feedback, please take a second look at the current state of this change

**Review by WilliamParker — APPROVED** [review]
I had a couple minor comments/thoughts but basically this looks good and I think it could be merged as is.

File: src/main/clojure/clara/rules/engine.cljc [hunk 8]
```diff
@@ -488,13 +488,15 @@
 
 (defn- throw-condition-exception
   "Adds a useful error message when executing a constraint node raises an exception."
-  [{:keys [cause node fact env bindings]}]
+  [{:keys [cause node fact env bindings] :as args}]
   (let [bindings-description (if (empty? bindings)
-                               "with no bindings\n"
+                               "with no bindings"
                                (str "with bindings\n  " bindings))
+        facts-description (if-not (contains? args :fact)
```
[the remark this answers is not in the document]
Comment (k13gomez) [reply src/main/clojure/clara/rules/engine.cljc:495]: good suggestion, makes sense, I think i will do that.

File: src/test/clojure/clara/test_rules.clj [hunk 9]
```diff
@@ -2361,10 +2361,31 @@
...
+                    (insert (->WindSpeed 10 "MCI") (->Temperature nil "MCI"))
+                    (fire-rules)
+                    (query check-exception))
+                (is false "Scope binding test is expected to throw an exception.")
+                (catch Exception e e))]
+        (is (= {:bindings {:?w 10, :?t nil}
+                :fact nil
+                :env {}
+                :conditions-and-rules {[:test '(> ?t ?w)] #{[:query "my-test-query"]}}}
+               (-> e (ex-data))))
```
[the remark this answers is not in the document]
Comment (k13gomez) [reply src/test/clojure/clara/test_rules.clj:2385]: It's not too difficult to make assert-ex-data also be able to assert the message, as an additional optional parameter, I'm going to make that change and submit it, it will make it easier to test both message and ex-data of nested exceptions in the future.

[the remark this answers is not in the document]
Comment (k13gomez) [reply src/test/clojure/clara/test_rules.clj]: @WilliamParker i've made a small change to `assert-ex-data` to allow checking the message along with the data as well, and updated the existing test to use the new `(assert-ex-data "message..." {...ex-data...} form)`.

[the remark this answers is not in the document]
Comment (WilliamParker) [reply src/test/clojure/clara/test_rules.clj]: +1 to those changes.
